akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -984,31 +997,176 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
   /**
    * Obtain a callback object that can be used asynchronously to re-enter the
    * current [[GraphStage]] with an asynchronous notification. The [[invoke()]] method of the returned
-   * [[AsyncCallback]] is safe to be called from other threads and it will in the background thread-safely
-   * delegate to the passed callback function. I.e. [[invoke()]] will be called by the external world and
+   * [[AsyncCallback]] is safe to be called from other threads. It will in the background thread-safely
+   * delegate to the passed callback function. I.e. [[invoke()]] will be called by other thread and
    * the passed handler will be invoked eventually in a thread-safe way by the execution environment.
    *
+   * In case stream is not yet materialized [[AsyncCallback]] will buffer events until stream is available.
+   *
+   * [[AsyncCallback.invokeWithFeedback()]] has an internal promise that will be failed if event cannot be processed
+   * due to stream completion.
+   *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
   final def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
-    new AsyncCallback[T] {
-      override def invoke(event: T): Unit =
-        interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
+    val result = new ConcurrentAsyncCallback[T](handler)
+    concurrentAsyncCallback.add(result)
+    if (_interpreter != null) result.onStart()
+    result
+  }
+
+  /**
+    * ConcurrentAsyncCallback allows to call [[invoke()]] and [[invokeWithPromise()]] with event attribute.
+    * This event will be sent to the stream and handler will be called with this attribute in thread safe manner.
+    *
+    * Class state can be changed by [[GraphStage]] and from "Real World" (other non Akka threads)
+    * * [[GraphStage]] calls [[onStart()]] - when materialization is finished and [[onStop()]] - when processing of external events stopped
+    *   because the stage is about to stop or fail.
+    * * "Real world" calls [[invoke()]] and [[invokeWithFeedback()]]. These methods have synchronization
+    *   with class state that reflects the stream state
+    *
+    * onStart sends all events that were buffered while stream was materializing.
+    * In case "Real world" added more events while initializing, onStart checks for more events in buffer when exiting and
+    * resend new events
+    *
+    * Once class is in `Initialized` state - all "Real world" calls of [[invoke()]] and [[invokeWithFeedback()]] are running
+    * as is - without any synchronization.
+    *
+    * [[GraphStage]] is called [[onStop()]] when stream is wrapping down. onStop fails all futures for events that have not yet processed
+    * [[onStop()]] puts class in `Completed` state
+    * "Real world" calls of [[invokeWithFeedback()]] always return failed promises for `Completed` state
+    */
+  private class ConcurrentAsyncCallback[T](handler: T ⇒ Unit) extends AsyncCallback[T] {
```

Comment:
I think this is correct, but thorough review by more people of this concurrent queuing is of importance